File: lib/features/game/domain/engine/baloot_deck.dart
import 'dart:math';

import 'package:bloot/features/game/domain/engine/baloot_card.dart';

/// Represents the 32-card Baloot deck and dealing operations.
class BalootDeck {
  BalootDeck({Random? random}) : _random = random ?? Random();

  final Random _random;

  /// Creates a fresh ordered 32-card deck.
  static List<BalootCard> makeDeck() {
    final deck = <BalootCard>[];
    for (final suit in BalootSuit.values) {
      for (final rank in BalootRank.values) {
        deck.add(BalootCard(suit: suit, rank: rank));
      }
    }
    return deck;
  }

  /// Fisher-Yates shuffle.
  List<BalootCard> shuffle(List<BalootCard> cards) {
    final result = List<BalootCard>.from(cards);
    for (var i = result.length - 1; i > 0; i--) {
      final j = _random.nextInt(i + 1);
      final tmp = result[i];
      result[i] = result[j];
      result[j] = tmp;
    }
    return result;
  }

  /// Returns a shuffled 32-card deck.
  List<BalootCard> shuffledDeck() => shuffle(makeDeck());
}

/// Result of a Baloot deal.
class BalootDealResult {
  const BalootDealResult({
    required this.hands,
    required this.topCard,
    required this.rest,
    required this.firstPlayer,
  });

  /// Four hands of 8 cards each after the full deal.
  final List<List<BalootCard>> hands;

  /// The face-up card that determines the proposed trump / first-round Hokm.
  final BalootCard topCard;

  /// The remaining 11 cards dealt in phase 2 (used internally by the engine).
  final List<BalootCard> rest;

  /// Seat index of the first player to bid / play.
  final int firstPlayer;
}

/// Performs the Saudi Baloot 2-phase deal.
///
/// Phase 1: 5 cards to each player.
/// Face-up card is revealed.
/// Phase 2: remaining 11 cards distributed so each player ends with 8.
BalootDealResult dealBalootHands({
  required BalootDeck deck,
  required int dealer,
}) {
  final d = deck.shuffledDeck();
  final hands = List.generate(4, (_) => <BalootCard>[]);
  final firstPlayer = (dealer + 1) % 4;

  // Phase 1: 5 cards each.
  var di = 0;
  for (var round = 0; round < 5; round++) {
    for (var p = 0; p < 4; p++) {
      hands[(firstPlayer + p) % 4].add(d[di++]);
    }
  }

  final topCard = d[di++];
  final rest = d.sublist(di); // 11 cards

  return BalootDealResult(
    hands: hands,
    topCard: topCard,
    rest: rest,
    firstPlayer: firstPlayer,
  );
}

File: functions/src/engine/balootDeck.ts
import { BalootCard, BALOOT_RANKS, BALOOT_SUITS } from './balootCard';

export interface BalootDealResult {
  hands: BalootCard[][];
  topCard: BalootCard;
  rest: BalootCard[];
  firstPlayer: number;
}

export class BalootDeck {
  private rng: () => number;

  constructor(rng: () => number = Math.random) {
    this.rng = rng;
  }

  static makeDeck(): BalootCard[] {
    const deck: BalootCard[] = [];
    for (const suit of BALOOT_SUITS) {
      for (const rank of BALOOT_RANKS) {
        deck.push(new BalootCard(suit, rank));
      }
    }
    return deck;
  }

  shuffle<T>(cards: T[]): T[] {
    const result = [...cards];
    for (let i = result.length - 1; i > 0; i--) {
      const j = Math.floor(this.rng() * (i + 1));
      [result[i], result[j]] = [result[j], result[i]];
    }
    return result;
  }

  shuffledDeck(): BalootCard[] {
    return this.shuffle(BalootDeck.makeDeck());
  }
}

/**
 * Performs the Saudi Baloot 2-phase deal:
 * - Phase 1: 5 cards to each player.
 * - Face-up card is revealed.
 * - Phase 2: remaining 11 cards distributed so each player ends with 8.
 */
export function dealBalootHands(deck: BalootDeck, dealer: number): BalootDealResult {
  const d = deck.shuffledDeck();
  const hands: BalootCard[][] = [[], [], [], []];
  // Bidding starts with the dealer (not the player to the dealer's right), per
  // the house rule set: the dealer is the first to choose Sun/Hokm/Pass.
  const firstPlayer = dealer;

  let di = 0;
  for (let round = 0; round < 5; round++) {
    for (let p = 0; p < 4; p++) {
      hands[(firstPlayer + p) % 4].push(d[di++]);
    }
  }

  const topCard = d[di++];
  const rest = d.slice(di); // 11 cards

  return { hands, topCard, rest, firstPlayer };
}
